### PR TITLE
WIP: Allow roles to be specified for Keycloak

### PR DESCRIPTION
**Pull request by khawaga**
## Description

This PR adds the ability to specify specific roles for use with Keycloak, which offers more granular control compared to groups.

## Motivation and Context

The motivation behind this feature was that a similar project, [Louketo Proxy](https://github.com/louketo/louketo-proxy), [has reached EOL](https://www.keycloak.org/2020/08/sunsetting-louketo-project.adoc.html), and that project had the ability to specify roles which was important for several applications. OAuth2 Proxy would be a suitable replacement if it were to offer that same feature.

## How Has This Been Tested?

The environment consisted of a Kubernetes cluster with a live Keycloak server.
I have tested the following:
- Not providing the new `keycloak-roles` flag to ensure that previous functionality is maintained. 
- Providing a role that my test user did not have. User was not granted access.
- Providing a realm role that my test user had. User was granted access.
- Providing a list consisting of two roles, one of which my test user did not have. User was not granted access. This is the desired behaviour as all roles need to exist in order for a user to be granted access.
- Providing a client role that my test user had. User was granted access.

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

File: contrib/oauth2-proxy_autocomplete.sh [hunk 1]
```diff
@@ -24,7 +24,7 @@ _oauth2_proxy() {
...
+		--@(http-address|https-address|redirect-url|upstream|basic-auth-password|skip-auth-regex|flush-interval|extra-jwt-issuers|email-domain|whitelist-domain|trusted-ip|keycloak-group|keycloak-roles|azure-tenant|bitbucket-team|bitbucket-repository|github-org|github-team|github-repo|github-token|gitlab-group|github-user|google-group|google-admin-email|google-service-account-json|client-id|client_secret|banner|footer|proxy-prefix|ping-path|cookie-name|cookie-secret|cookie-domain|cookie-path|cookie-expire|cookie-refresh|cookie-samesite|redist-sentinel-master-name|redist-sentinel-connection-urls|redist-cluster-connection-urls|logging-max-size|logging-max-age|logging-max-backups|standard-logging-format|request-logging-format|exclude-logging-paths|auth-logging-format|oidc-issuer-url|oidc-jwks-url|login-url|redeem-url|profile-url|resource|validate-url|scope|approval-prompt|signature-key|acr-values|jwt-key|pubjwk-url))
```
Comment (NickMeves) [inline contrib/oauth2-proxy_autocomplete.sh:27]: Ha - I didn't even realize this was a thing.  I bet this way out of date, I've seen many PRs adding flags that didn't touch this 🤣

File: providers/keycloak.go [hunk 2]
```diff
@@ -95,5 +128,29 @@ func (p *KeycloakProvider) GetEmailAddress(ctx context.Context, s *sessions.Sess
 		}
 	}
 
+	if len(p.Roles) > 0 {
```
Comment (NickMeves) [inline providers/keycloak.go:131]: Can you refactor this into another method that does the role authorization? This addition makes this `GetEmailAddress` method too complex: https://codeclimate.com/github/oauth2-proxy/oauth2-proxy/pull/767

File: providers/keycloak.go [hunk 3]
```diff
@@ -95,5 +128,29 @@ func (p *KeycloakProvider) GetEmailAddress(ctx context.Context, s *sessions.Sess
...
+		}
+
+		// Parse claims
+		if err := token.UnsafeClaimsWithoutVerification(&claims); err != nil {
+			logger.Printf("failed to parse claims %s", err)
+		}
+
+		var roles = ExtractRolesFromClaims(claims)
+
+		if !isSubarray(roles, p.Roles) {
```
Comment (NickMeves) [inline providers/keycloak.go:149]: So we might need to think about how this aligns with other implementations. In general, in other group/role AuthZ logic in other providers, we've made this a if any roles match the list, access is allowed (instead of all must be present).

File: providers/keycloak.go [hunk 4]
```diff
@@ -63,12 +66,42 @@ func (p *KeycloakProvider) SetGroup(group string) {
 	p.Group = group
 }
 
+func (p *KeycloakProvider) SetRoles(roles []string) {
+	p.Roles = roles
+}
+
+func ExtractRolesFromClaims(claims map[string]interface{}) []string {
```
Comment (NickMeves) [inline providers/keycloak.go:73]: Can you add Unit Tests for this new method? (And add new cases for the additions you made to existing methods)

**Review by NickMeves — CHANGES_REQUESTED** [review]
Thanks for submitting this!  I left some notes to try to get this aligned with the other group/role work that is ongoing in other providers.

File: providers/keycloak.go (hunk 4 above)
Comment (khawaga) [reply providers/keycloak.go:73]: Sure thing, would it make sense to change [auth_test.go](https://github.com/oauth2-proxy/oauth2-proxy/blob/master/providers/auth_test.go#L13) in order to allow passing an access token when calling `CreateAuthorizedSession`? Since the roles are extracted from the access token so it needs to be valid otherwise token parsing will fail.

File: contrib/oauth2-proxy_autocomplete.sh (hunk 1 above)
Comment (khawaga) [reply contrib/oauth2-proxy_autocomplete.sh:27]: I did not know that 😂, I just did a search for one of the existing flags and that's how I found this! :)

File: providers/keycloak.go (hunk 2 above)
Comment (khawaga) [reply providers/keycloak.go:131]: Sure thing, but I believe after refactoring block into its own method it would still exceed the number of allowed returns (it would become 1 less, leaving it at 5, which still exceeds the 4 allowed) so that would not fix the issue. Would you still suggest to refactor it?

File: providers/keycloak.go (hunk 3 above)
Comment (khawaga) [reply providers/keycloak.go:149]: Oh, are roles being used in any other provider? The only thing I've seen is groups, and for that it makes sense that access would be granted for a user belonging to any specified group. This was based on the Louketo Proxy implementation.

> By default, the roles defined on a resource perform a logical AND so all roles specified must be present in the claims, this behavior can be altered by the require-any-role option, however, so as long as one role is present the permission is granted.

I am open for suggestions though, would you like me to change it so that it grants access if any of the provided roles are present?

File: providers/keycloak.go (hunk 4 above)
Comment (khawaga) [reply providers/keycloak.go:73]: I've added two more tests.

Comment (khawaga) [reply providers/keycloak.go:73]: I've edited the token in auth_test.go in order to contain a valid token instead of the dummy that was being used. It's currently hardcoded which is not exactly ideal, but if you'd like I can add another method to create the token (that can then be called from the tests in order to create and then set it).

**Comment by khawaga** [comment]
@NickMeves I've added some tests and I've also replied to your comments. Thank you for the prompt review!

File: providers/auth_test.go [hunk 5]
```diff
@@ -8,7 +8,8 @@ import (
 	"github.com/oauth2-proxy/oauth2-proxy/pkg/apis/sessions"
 )
 
-var authorizedAccessToken = "imaginary_access_token"
+// The following is a valid JWT, it can be decoded, edited and re-encoded using https://jwt.io/
+var authorizedAccessToken = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkhvbWVyIFNpbXBzb24iLCJlbWFpbCI6IkNodW5reUxvdmVyNTNAYW9sLmNvbSIsImlhdCI6MTUxNjIzOTAyMiwicmVhbG1fYWNjZXNzIjp7InJvbGVzIjpbInRlc3QtcmVhbG1yb2xlMSIsInRlc3QtcmVhbG1yb2xlMiJdfSwicmVzb3VyY2VfYWNjZXNzIjp7ImNsaWVudCI6eyJyb2xlcyI6WyJ0ZXN0LWNsaWVudHJvbGUxIiwidGVzdC1jbGllbnRyb2xlMiJdfX19.DqKfcantBn7B8acd9rl0LK8FL3sVhUcrM_AHztWI2A0"
```
Comment (JoelSpeed) [inline providers/auth_test.go:12]: What key was used to sign this?

File: providers/keycloak.go (hunk 4 above)
Comment (JoelSpeed) [reply providers/keycloak.go:73]: Does this function need to be exported?

File: providers/keycloak.go [hunk 6]
```diff
@@ -63,12 +66,42 @@ func (p *KeycloakProvider) SetGroup(group string) {
...
+	return roleList
+}
+
 func (p *KeycloakProvider) GetEmailAddress(ctx context.Context, s *sessions.SessionState) (string, error) {
 	json, err := requests.New(p.ValidateURL.String()).
 		WithContext(ctx).
 		SetHeader("Authorization", "Bearer "+s.AccessToken).
 		Do().
 		UnmarshalJSON()
+
```
Comment (JoelSpeed) [inline providers/keycloak.go:104]: Any reason for this new line?

File: providers/keycloak.go (hunk 2 above)
Comment (JoelSpeed) [reply providers/keycloak.go:131]: Code climate is advisory, take it with a pinch of salt. Would be good still to have a dedicated role check function so basically just extract the contents of this if to a function I would say

File: providers/keycloak.go [hunk 7]
```diff
@@ -95,5 +128,29 @@ func (p *KeycloakProvider) GetEmailAddress(ctx context.Context, s *sessions.Sess
 		}
 	}
 
+	if len(p.Roles) > 0 {
+		claims := make(map[string]interface{})
+
+		// Decode JWT token without verifying the signature
+		token, err := jwt.ParseSigned(s.AccessToken)
+
+		if err != nil {
```
Comment (JoelSpeed) [inline providers/keycloak.go:137]: Would prefer not to have a new line between the err declaration and check if the err happened

File: providers/keycloak.go (hunk 3 above)
Comment (JoelSpeed) [inline providers/keycloak.go:149]: Nit
```suggestion
		roles := ExtractRolesFromClaims(claims)
		if !isSubarray(roles, p.Roles) {
```

File: providers/util.go [hunk 8]
```diff
@@ -29,3 +29,20 @@ func makeOIDCHeader(accessToken string) http.Header {
 	}
 	return makeAuthorizationHeader(tokenTypeBearer, accessToken, extraHeaders)
 }
+
+// Return true if secondArray is a subset of firstArray
+func isSubarray(firstArray, secondArray []string) bool {
+	arraySet := make(map[string]bool)
```
Comment (JoelSpeed) [inline providers/util.go:35]: Nit, better to use `map[string]struct{}` and then `arraySet[value] = struct{}{}` when setting it and `if _, found := arraySet[value]; !found` when checking, this is more explicit and idiomatic in Go.

Comment (khawaga) [reply providers/util.go]: I truly appreciate the tip! I changed it now.

File: providers/keycloak.go (hunk 7 above)
Comment (khawaga) [reply providers/keycloak.go]: Got it!

File: providers/keycloak.go (hunk 6 above)
Comment (khawaga) [reply providers/keycloak.go]: I removed it.

File: providers/auth_test.go (hunk 5 above)
Comment (khawaga) [reply providers/auth_test.go:12]: It's using the default key on jwt.io which is "your-256-bit-secret" 😅
Should I change it to something else? Or add it as a comment?

File: providers/keycloak.go (hunk 4 above)
Comment (khawaga) [reply providers/keycloak.go]: It does not, I unexported it :)

File: providers/keycloak.go (hunk 2 above)
Comment (khawaga) [reply providers/keycloak.go:146]: Got it, I've extracted the logic into a separate function.

**Comment by khawaga** [comment]
@NickMeves I've done the requested changes and I've also replied to your comments, thanks!

File: providers/keycloak.go (hunk 3 above)
Comment (khawaga) [reply providers/keycloak.go]: Thank you for the suggestion, I changed it :)

**Review by NickMeves — CHANGES_REQUESTED** [review]
Sorry if these comments didn't get submitted before... I guess I just left them pending somehow and didn't get them over to you. Forgive me if some are outdated at this point -- but I see a few issues that might still need to be resolved.

**Comment by khawaga** [comment]
@NickMeves I've changed the logic (and the docs) to use ANY instead of ALL for role-matching, and I've also added some OK checks to the type assertions. Are there any other issues that need to be resolved?

**Comment by NickMeves** [comment]
Adding in some minor scope creep:

PR #616 which just merged did very similar AuthZ work as this for the OIDC provider.  It additionally added in a global `Groups` field to all sessions that when set would result in `X-Forwarded-Groups` HTTP header to backend applications (so they could use it for fine-grained route based RBAC themselves):

https://github.com/oauth2-proxy/oauth2-proxy/pull/616/files#diff-6fd8df33d9f8086bc31e28c375fbc0abR951

I think we should dump the roles you extract into the session so users get access to these details in their backend applications easily.

I know the names don't fully line up, but the bazillion providers all have different group-like concepts with different names -- so I think `Groups` is the ideal area to centralize them letting our first class provider OIDC get the naming that mirrors its conventions.

**Comment by NickMeves** [comment]
> @NickMeves I've changed the logic (and the docs) to use ANY instead of ALL for role-matching, and I've also added some ok checks to the type assertions. Are there any other issues that need to be resolved?

I'll give everything another look this afternoon.  For now just take care of the code climate findings: https://codeclimate.com/github/oauth2-proxy/oauth2-proxy/pull/767

Are these the same findings we talked about before that might've been challenging to refactor? Or are these new?

**Comment by NickMeves** [comment]
So I did some work this weekend to cleanup the session enrichment process after token redeem (https://github.com/oauth2-proxy/oauth2-proxy/pull/796) & centralizing authorization logic to be more global and uniform across providers (https://github.com/oauth2-proxy/oauth2-proxy/pull/797) since we are having so many authorization PRs lately taking their providers in different directions.

This PR caught my eye since it looks like the previous Keycloak roles logic was just tossed into `GetEmailAddress` since no better place existed even though it ideally shouldn't be there.  The authZ logic also looks very similar to other places, so hopefully it can just hook into the default `Authorize` provider method (or implement its own and have it called at the right moments by `oauthproxy`.

I'll check back once those PRs merge.

**Comment by NickMeves** [comment]
Sorry for the delay.  If you still want to get this over the finish line:

https://github.com/oauth2-proxy/oauth2-proxy/pull/797 merged.

The refactor to `EnrichSessionState` & `Authorize` provider interface methods are now in master. This should be completely unblocked now.

**Comment by khawaga** [comment]
@NickMeves I would still like to see this through, definitely.
I'll take a look at the required changes and refactor. Thanks for the heads-up!

File: CHANGELOG.md [hunk 9]
```diff
@@ -12,6 +12,7 @@
 
 - [#764](https://github.com/oauth2-proxy/oauth2-proxy/pull/764) Document bcrypt encryption for htpasswd (and hide SHA) (@lentzi90)
 - [#616](https://github.com/oauth2-proxy/oauth2-proxy/pull/616) Add support to ensure user belongs in required groups when using the OIDC provider (@stefansedich)
+- [#757](https://github.com/oauth2-proxy/oauth2-proxy/pull/767) Add ability to specify roles when using the Keycloak provider (@khawaga)
```
Comment (bakito) [inline CHANGELOG.md:15]: You have a typo here. Should be #767

**Comment by NickMeves** [comment]
As a heads up, we are likely looking to make a V7 release sometime in mid-January.  Just wanted to give you a timeline update in case you were looking to get this in for the next release.

File: CHANGELOG.md (hunk 9 above)
Comment (khawaga) [reply CHANGELOG.md]: Thanks for the heads-up! I fixed it!

File: pkg/apis/options/options.go [hunk 10]
```diff
@@ -37,6 +37,7 @@ type Options struct {
 
 	AuthenticatedEmailsFile  string   `flag:"authenticated-emails-file" cfg:"authenticated_emails_file"`
 	KeycloakGroup            string   `flag:"keycloak-group" cfg:"keycloak_group"`
+	KeycloakRoles            []string `flag:"keycloak-roles" cfg:"keycloak_roles"`
```
Comment (NickMeves) [inline pkg/apis/options/options.go:40]:
```suggestion
	KeycloakRoles            []string `flag:"keycloak-role" cfg:"keycloak_roles"`
```

Flags are singular, configs are plural in multi-param cases.

File: pkg/apis/sessions/session_state.go [hunk 11]
```diff
@@ -27,6 +27,7 @@ type SessionState struct {
 	Email             string   `msgpack:"e,omitempty"`
 	User              string   `msgpack:"u,omitempty"`
 	Groups            []string `msgpack:"g,omitempty"`
+	Roles             []string `msgpack:"r,omitempty"`
```
Comment (NickMeves) [inline pkg/apis/sessions/session_state.go:30]: Nothing new should be added to the session. You should add the roles to the existing `session.Groups` option to ride default `Authorize` & header injection implementations. This helps keep the codebase more uniform so providers can automatically get new features as we adopt them globally if they adhere to conventions.

When providers have multiple group-like structures that might want to be in this field, we've adopted prefixing the names.  Here's a sample from a very similar GitLab PR that just merged: https://github.com/oauth2-proxy/oauth2-proxy/pull/630/files#diff-9179d3c8d617839cfc99bbcd5a63a315b6cf3e652bb7f7fb9dd4bd8c6b40128aR310

File: providers/auth_test.go (hunk 5 above)
Comment (NickMeves) [reply providers/auth_test.go:12]: I'm centralizing OIDC logic in this PR and making sample IDTokens more accessible to test cases across providers here: https://github.com/oauth2-proxy/oauth2-proxy/pull/936/files#diff-e719a8f54d2fe3115f6fcaabd056be5b53e86a3f04004d3c53f0b6c5b0409d68R34

Might be of use to this PR if you need valid IDTokens for testing.

File: providers/keycloak.go [hunk 12]
```diff
@@ -63,37 +68,151 @@ func (p *KeycloakProvider) SetGroup(group string) {
...
+		logger.Printf("failed to parse claims: %s", err)
+	}
+	roles, err := extractRolesFromClaims(claims)
+	if err != nil {
+		logger.Printf("failed to extract roles: %s", err)
+	}
+	return roles, nil
+}
+
+func (p *KeycloakProvider) isAuthorizedGroupMember(s *sessions.SessionState) (bool, error) {
```
Comment (NickMeves) [inline providers/keycloak.go:157]: This should be unneeded if you set the roles you want like this to take advantage of the global `Authorize` logic. Use a prefixed name like I mentioned above to not collide with groups -- e.g. `role:foorbar`

https://github.com/oauth2-proxy/oauth2-proxy/blob/d67d6e3152e2a1c9614906ee9bd9890f8da9a007/pkg/validation/options.go#L289

Here are the pertinent areas that do this for you:
https://github.com/oauth2-proxy/oauth2-proxy/blob/d67d6e3152e2a1c9614906ee9bd9890f8da9a007/providers/provider_data.go#L56

https://github.com/oauth2-proxy/oauth2-proxy/blob/d67d6e3152e2a1c9614906ee9bd9890f8da9a007/providers/provider_default.go#L102

File: providers/keycloak.go [hunk 13]
```diff
@@ -63,37 +68,151 @@ func (p *KeycloakProvider) SetGroup(group string) {
...
 		if !found {
-			logger.Printf("group not found, access denied")
-			return "", nil
+			return false, errors.New("group not found, access denied")
+		}
+	}
+	return true, nil
+}
+
+func (p *KeycloakProvider) hasRequiredRole(s *sessions.SessionState) (bool, error) {
```
Comment (NickMeves) [inline providers/keycloak.go:173]: This is also unneeded in favor of the new `SetAllowedGroups` & `Authorize` functionality

File: providers/keycloak.go [hunk 14]
```diff
@@ -63,37 +68,151 @@ func (p *KeycloakProvider) SetGroup(group string) {
...
-		var groups, err = json.Get("groups").Array()
+func getEmailAddress(userInfo *simplejson.Json) (string, error) {
+	var email, err = userInfo.Get("email").String()
+	if err != nil {
+		return "", fmt.Errorf("error getting email address: %v", err)
+	}
+	return email, nil
+}
+
+func getGroups(userInfo *simplejson.Json) ([]string, error) {
```
Comment (NickMeves) [inline providers/keycloak.go:128]: Will we need to harmonize this groups scope creep with this related PR if it gets picked up again? https://github.com/oauth2-proxy/oauth2-proxy/pull/861

File: providers/keycloak.go [hunk 15]
```diff
@@ -63,37 +68,151 @@ func (p *KeycloakProvider) SetGroup(group string) {
...
 
+func getRoles(s *sessions.SessionState) ([]string, error) {
+	claims := make(map[string]interface{})
+	// Decode JWT token without verifying the signature
+	token, err := jwt.ParseSigned(s.AccessToken)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse token: %v", err)
+	}
+	// Parse claims
+	if err := token.UnsafeClaimsWithoutVerification(&claims); err != nil {
```
Comment (NickMeves) [inline providers/keycloak.go:147]: You've gotta `Verify` tokens if we want to use any data from them as part of the session. I'm gonna be a stickler for that for security reasons.

We've made `Verifier *oidc.IDTokenVerifier` an object available globally that any providers that set it can use.
https://github.com/oauth2-proxy/oauth2-proxy/blob/d67d6e3152e2a1c9614906ee9bd9890f8da9a007/pkg/validation/options.go#L237

Take a look at this PR and maybe wait until it merges to see if its new global OIDC helpers make sense for Keycloak and how you extract claims: https://github.com/oauth2-proxy/oauth2-proxy/pull/936

Keycloak, Gitlab & Azure all have OIDC elements, that PR's goal is to unify the logic in how we turn IDToken claims into Session fields.

File: providers/keycloak.go [hunk 16]
```diff
@@ -63,37 +68,151 @@ func (p *KeycloakProvider) SetGroup(group string) {
...
-func (p *KeycloakProvider) GetEmailAddress(ctx context.Context, s *sessions.SessionState) (string, error) {
-	json, err := requests.New(p.ValidateURL.String()).
+func (p *KeycloakProvider) SetRoles(roles []string) {
+	p.Roles = roles
+}
+
+func extractRolesFromClaims(claims map[string]interface{}) ([]string, error) {
+	var roleList []string
+
+	if realmRoles, found := claims["realm_access"].(map[string]interface{}); found {
```
Comment (NickMeves) [inline providers/keycloak.go:78]: By chance do you have a sample of what Keycloak IDToken claims are present? I'm curious if some of the data pulled via the userinfo/profileURL endpoint can potentially be extracted from IDToken claims and we don't need that extra HTTP call.

Comment (NickMeves) [reply providers/keycloak.go:78]: (This is definitely scope creep and not needed for your PR -- I'm just curious for my own understanding).

**Review by NickMeves — CHANGES_REQUESTED** [review]
So I think all my refactors since your originally submitted this have changed the lay of the land a bit 😅 

I've tried to leave notes on how to integrate with the new standardized/global ways we've setup to do this. Please ask whatever question you need, I'll try to respond quickly to get you going in the right direction with some of the new architectures.

I haven't looked at the new test cases at all -- I'll review those once we get the code aligned to the new standards.

File: pkg/apis/sessions/session_state.go (hunk 11 above)
Comment (khawaga) [reply pkg/apis/sessions/session_state.go:30]: Got it, I will make the required changes.

File: providers/keycloak.go (hunk 13 above)
Comment (khawaga) [reply providers/keycloak.go:173]: Got it, looking at that now that actually makes this much simpler and cleaner. One part I'd like to discuss here though, I've noticed that this would effectively result in an ANY behaviour between roles and groups which may not fit here, I would appreciate your input on this. I've seen that this was already discussed here, although in a different context: https://github.com/oauth2-proxy/oauth2-proxy/pull/630/files#r529172740

File: providers/keycloak.go (hunk 16 above)
Comment (khawaga) [reply providers/keycloak.go:78]: Not sure about the IDToken, but the AccessToken contains both the email and the groups which should be sufficient to get the info used here. I would say that HTTP call is not particularly necessary.

Comment (khawaga) [reply providers/keycloak.go:78]: Although that might require a breaking change since I believe that's configurable, if groups are returned via userinfo but not included in the token.

Comment (khawaga) [reply providers/keycloak.go:78]: I checked, and it should possible to get the required info from the IDToken as well with the right configuration.

File: providers/keycloak.go (hunk 14 above)
Comment (khawaga) [reply providers/keycloak.go]: I believe so.

File: providers/keycloak.go (hunk 12 above)
Comment (khawaga) [reply providers/keycloak.go]: I've done that, I would still appreciate your input on my comment regarding the groups/roles ANY logic on the other thread.

File: providers/keycloak.go (hunk 13 above)
Comment (NickMeves) [reply providers/keycloak.go]: For the case where `--keycloak-groups` & `--keycloak-roles` is set, if you want to enforce that a user has at least 1 group AND at least 1 role that matches, you should override the default `Authorize` method in the Keycloak provider.

I'd imagine you'd manage 2 sets (groups and roles) and force membership checks in BOTH sets to return true to result in `Authorize` returning true.

File: providers/keycloak.go (hunk 16 above)
Comment (NickMeves) [reply providers/keycloak.go]: Oh interesting... that would definitely make this different than other providers that are more OIDC based and get extra claims from an IDToken we can verify.

Does Keycloak have any documentation on where to get the public key to verify the signature on the Access Tokens?

I really only want to use claims extracted from an unverified token as a last resort. There's gotta be a way we can verify the signature of the token?

File: providers/keycloak.go [hunk 17]
```diff
@@ -63,37 +67,126 @@ func (p *KeycloakProvider) SetGroup(group string) {
...
+	customClaims := customClaims{}
+	// Parse claims
+	if err := token.UnsafeClaimsWithoutVerification(&standardClaims, &customClaims); err != nil {
+		logger.Printf("failed to parse claims: %s", err)
+	}
+	clientRoles, err := getClientRoles(customClaims.ResourceAccess)
+	if err != nil {
+		logger.Printf("failed to extract client roles: %s", err)
+	}
+	return append(customClaims.RealmAccess.Roles, clientRoles...), nil
```
Comment (NickMeves) [inline providers/keycloak.go:134]: It feels weird to pollute & extend the underlying object we extracted from the token claims.

I understand the performance motivations of reusing the existing slice since it is done at the end of the function anyway, but since these are small & therefore performance is negligible, maybe clarity is more ideal? Took me a second on first glance to realize the desired output was intended as `RealmAccess` + ResourceAccess`

File: providers/keycloak.go [hunk 18]
```diff
@@ -63,37 +67,126 @@ func (p *KeycloakProvider) SetGroup(group string) {
...
+	}
+	return &userInfo, nil
+}
+
+func getClientRoles(resourceAccess map[string]interface{}) ([]string, error) {
+	var clientRoles []string
+	for name, list := range resourceAccess {
+		scopes, ok := list.(map[string]interface{})
+		if !ok {
+			return nil, errors.New("error parsing client roles from claims")
```
Comment (NickMeves) [inline providers/keycloak.go:98]: Do you want to completely error if only 1 element in the list had parsing errors or skip it and make a log message?

OIDC complex claims code went the second route: https://github.com/oauth2-proxy/oauth2-proxy/blob/d67d6e3152e2a1c9614906ee9bd9890f8da9a007/providers/oidc.go#L294

File: providers/keycloak.go [hunk 19]
```diff
@@ -63,37 +67,126 @@ func (p *KeycloakProvider) SetGroup(group string) {
...
+		}
+		if roles, found := scopes["roles"]; found {
+			for _, r := range roles.([]interface{}) {
```
Comment (NickMeves) [inline providers/keycloak.go:101]: Do you need to handle when this isn't a `[]interface{}` or have fallback code that somehow marshals non-slices if you get a singleton? We have to handle that case in the OIDC side: https://github.com/oauth2-proxy/oauth2-proxy/pull/936/commits/48baecb5c996d378f6871fa41a1c4824354dd083

File: providers/keycloak.go [hunk 20]
```diff
@@ -63,37 +67,126 @@ func (p *KeycloakProvider) SetGroup(group string) {
...
+				clientRoles = append(clientRoles, fmt.Sprintf("%s:%s", name, r))
```
Comment (NickMeves) [inline providers/keycloak.go:102]: Will `name` always be `role`? Should we make that prefix explicit here instead of from a variable so we don't get surprised later?

Comment (NickMeves) [reply providers/keycloak.go:102]: Oh wait, nevermind - I see the `role` prefix later.  So this has a different `name:role` structure that represents the full role name?

File: providers/keycloak.go (hunk 16 above)
Comment (NickMeves) [reply providers/keycloak.go]: Is this an option? https://stackoverflow.com/questions/54318633/getting-keycloaks-public-key

File: providers/keycloak.go (hunk 20 above)
Comment (khawaga) [reply providers/keycloak.go:102]: Yes, so effectively this would end up in allowed groups as role:clientname:clientrole. These are client roles which are namespaced by client, so while the structure is predictable the client name could be anything, hence the additional prefix.

File: providers/keycloak.go (hunk 17 above)
Comment (khawaga) [reply providers/keycloak.go:134]: You're right, I will change this for better clarity.

File: providers/keycloak.go (hunk 16 above)
Comment (khawaga) [reply providers/keycloak.go]: Keycloak has a JWKS endpoint at `/auth/realms/<realm_name>/protocol/openid-connect/certs`, it should be possible to verify it using that, should I reuse `oidc-jwks-url` for this purpose? If so, what would be the best way to pass it to the provider?

File: providers/keycloak.go (hunk 13 above)
Comment (khawaga) [reply providers/keycloak.go]: I think that would make more sense, what do you think? Should I go ahead and do that then?

Comment (NickMeves) [reply providers/keycloak.go]: Yeah, if you need separate ANY logic between groups and roles, and between the sets AND logic, overriding the default with a custom `Authorize` method that does that logic is best.

File: providers/keycloak.go (hunk 16 above)
Comment (NickMeves) [reply providers/keycloak.go]: So I need your help explaining something 😅 

Keycloak as a provider confuses me. When I look at it, it has barebones implementations as if it were vanilla OAuth2. (and should what it calls the `validateURL` be the `profileURL`...).

BUT - I get sample configs from users using OIDC config options with it set as the provider in questions frequently. If it is truly OIDC compliant in all cases, I feel like it should be baked into the provider more.

This way we can take advantage of all the new OIDC based features we are making global and Keycloak can take advantage of them (bearer token to session logic, generic claims to session fields - future adhoc claims -> session -> header support).

The reason Keycloak would still be a distinct provider instead of just using Keycloak with an OIDC provider: your custom roles from access tokens logic here, any special quirks in the schema of the userinfo endpoint etc.

Global OIDC logic refactor: https://github.com/oauth2-proxy/oauth2-proxy/pull/936

CC: @JoelSpeed -- if you have some history context or experience with this provider yourself, I would love your thoughts.

File: providers/keycloak.go (hunk 19 above)
Comment (khawaga) [reply providers/keycloak.go:101]: From what I've seen it always returns a slice even when if it only contains a single element, is there any other case that I'm missing?

File: providers/keycloak.go (hunk 16 above)
Comment (khawaga) [reply providers/keycloak.go]: My understanding is that it is OIDC-compliant (it's OpenID-certified: https://openid.net/developers/certified/), I believe validateURL should be userInfoURL.
It _should_ work with the standard OIDC provider implementation (although I have not tried that myself), but yeah basically the additional role functionality is probably worth having it as an independent provider.

File: providers/keycloak.go (hunk 19 above)
Comment (NickMeves) [reply providers/keycloak.go:101]: Cool - that makes sense that it is uniform since it is all provided by the singular Keycloak. OIDC is all over the place cause identity providers do some crazy stuff with `groups` since it isn't a totally official claim.

File: providers/keycloak.go (hunk 16 above)
Comment (NickMeves) [reply providers/keycloak.go]: So I think to get at an OIDC IDToken Verifier (if that works on the Access Token) or underlying raw JWKS, we use the existing flags to set it: https://oauth2-proxy.github.io/oauth2-proxy/docs/configuration/oauth_provider/#skip-oidc-discovery

During options validation, if you see `--keycloak-roles` you'd need to assert you also have the OIDC URLs setup so you have the tools at your disposal to verify the signature of the Access Token.

Comment (NickMeves) [reply providers/keycloak.go]: And on the userinfo... I guess technically it works as a validateURL since it will take an access token and return 200, which is all this is looking for: https://github.com/oauth2-proxy/oauth2-proxy/blob/d67d6e3152e2a1c9614906ee9bd9890f8da9a007/providers/internal_util.go#L49

But it is primarily used as a profileURL in the Keycloak provider itself.  I'll see if I can get a PR up to clean that up in a backwards compatible way.

**Comment by khawaga** [comment]
@NickMeves since #953 is now merged I'll rebase my branch on top of that, I'll post an update here once that's done.

**Comment by NickMeves** [comment]
> @NickMeves since #953 is now merged I'll rebase my branch on top of that, I'll post an update here once that's done.

So take a look at this issue: https://github.com/oauth2-proxy/oauth2-proxy/issues/956

Long term, there's a few providers that I want to move from being OAuth based to OIDC based to embrace all that functionality. Keycloak is one of them (especially with the features you want to add). The existing `api` scope OAuth provider sorta sucks and I've realized alot of Keycloak users are hooking it into the `oidc` provider (after getting burned by thinking `keycloak` provider is for them but is missing functionality).

So after working on the existing `keycloak` provider. I think it eventually needs to get deprecated in favor of an OIDC based variant.

I think we make a new `keycloak-oidc` provider for the time being (in v8 it will replace the `keycloak`). We can have it be based on the `OIDCProvider` as a base. Then I think you get nearly all the functionality you need in the OIDC methods and you just need to add a few methods that extracts the roles from the Access Token.

That also gets you the OIDC supporting methods to appropriately verify your Access Token before extracting claims from it.

What do you think?

Let me know if you need a hand spiking out what a base `keycloak-oidc` provider would look like.

**Comment by khawaga** [comment]
@NickMeves what you're suggesting makes perfect sense, is there currently any provider that's following this pattern (using the OIDC provider as a base, then extending it with additional functionality) that I can look at? I would appreciate your input on this. :)

**Comment by NickMeves** [comment]
> @NickMeves what you're suggesting makes perfect sense, is there currently any provider that's following this pattern (using the OIDC provider as a base, then extending it with additional functionality) that I can look at? I would appreciate your help on this. :)

@khawaga You would be the first - the trailblazer 😉 

Let me know if you need a hand. I can find time over the weekend to spike out a baseline `keycloak-oidc` provider that would completely inherit `OIDCProvider` methods (it would have `OIDCProvider` in its struct composition instead of `ProviderData` in essence).

Then from that you could add on the few overrides for things like `EnrichSession` to handle the custom Access Token role logic.

**Comment by NickMeves** [comment]
I will finish this work in #1107 since this PR has gone stale.
